core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -1294,6 +1330,22 @@ private[spark] class AppStatusListener(
       cleanupCachedQuantiles(key)
       key
     }
+  }
+
+  private def cleanupStages(count: Long): Unit = {
+    val countToDelete = calculateNumberToRemove(count, conf.get(MAX_RETAINED_STAGES))
+    if (countToDelete <= 0L) {
+      return
+    }
+
+    // SPARK-36827: For better performance and avoiding OOM, here we use a optimized method for
+    //              cleaning the StageDataWrapper and RDDOperationGraphWrapper data if Spark is
+    //              using InMemoryStore.
+    val stageIds = if (kvstore.usingInMemoryStore) {
+      cleanupStagesWithInMemoryStore(countToDelete)
```

Comment:
Could this be pushed down to the in memory store? vs special casing it here. This doesn't seem very object-oriented

Reply:
+1 for @srowen 's comment.

Reply:
maybe we need a more high-level API in the kv store to clean up stages?

Reply:
@srowen +1. I think we can have an enhanced version of in-memory `ElementTrackingStore`.
I was trying to merge this one in 3.2.0 so this is a temporary solution.